File: services/python-worker/worker_plugin_system.py
#!/usr/bin/env python3
"""
Python Worker with Plugin System
Auto-discovers and loads all plugins from plugins/ directory
"""

import grpc
import json
import time
import sys
import os
import uuid
import threading
import queue
import traceback
from datetime import datetime

# Import generated proto files
import hub_pb2
import hub_pb2_grpc

# Import plugin system
from plugin_manager import PluginManager


class PluginWorker:
    """Worker that uses plugin system for capability management"""
    
    def __init__(self, worker_id='python-worker', hub_address='localhost:50051'):
        self.worker_id = worker_id
        self.hub_address = hub_address
        self.channel = None
        self.stub = None
        self.running = False
        self.send_queue = None
        
        # Worker-to-worker call tracking
        self.pending_calls = {}
        self.pending_lock = threading.Lock()
        
        # Plugin system
        self.plugin_manager = PluginManager()
        self.plugins = {}
        
    def load_plugins(self):
        """Auto-discover and load all plugins"""
        self.plugins = self.plugin_manager.load_all_plugins()
        
        if not self.plugins:
            print("⚠️  No plugins loaded! Worker will have no capabilities.")
        
        return self.plugins
    
    def call_worker(self, target_worker, capability, params, timeout=30):
        """
        Call another worker's capability through Hub
        
        Args:
            target_worker: ID of the target worker
            capability: Capability name to call
            params: Dictionary of parameters
            timeout: Timeout in seconds
            
        Returns:
            Response data as dictionary
        """
        if not self.send_queue:
            raise RuntimeError("Worker not connected")
        
        request_id = str(uuid.uuid4())
        print(f"🔗 Calling worker '{target_worker}' capability '{capability}'")
        
        # Create WORKER_CALL message
        call_msg = hub_pb2.Message(
            id=request_id,
            to=target_worker,
            channel=capability,
            content=json.dumps(params),
            timestamp=datetime.now().isoformat(),
            type=hub_pb2.WORKER_CALL
        )
        setattr(call_msg, 'from', self.worker_id)
        call_msg.metadata['capability'] = capability
        
        # Register pending call
        response_event = threading.Event()
        response_data = {'response': None, 'error': None}
        
        with self.pending_lock:
            self.pending_calls[request_id] = {
                'event': response_event,
                'data': response_data
            }
        
        # Send the call
        self.send_queue.put(call_msg)
        
        # Wait for response
        if response_event.wait(timeout=timeout):
            with self.pending_lock:
                self.pending_calls.pop(request_id, None)
            
            if response_data['error']:
                raise Exception(response_data['error'])
            
            return response_data['response']
        else:
            with self.pending_lock:
                self.pending_calls.pop(request_id, None)
            raise TimeoutError(f"No response from {target_worker} after {timeout}s")
    
    def _handle_worker_call_response(self, msg):
        """Handle response from worker-to-worker call"""
        request_id = msg.metadata.get('request_id', '')
        
        with self.pending_lock:
            if request_id and request_id in self.pending_calls:
                call_info = self.pending_calls[request_id]
                try:
                    response_content = json.loads(msg.content)
                    call_info['data']['response'] = response_content
                    call_info['event'].set()
                except Exception as e:
                    call_info['data']['error'] = f"Failed to parse response: {e}"
                    call_info['event'].set()
    
    def process_message(self, msg):
        """Process incoming message using plugin system"""
        capability_name = msg.channel
        
        print(f"  → Processing capability: {capability_name}")
        
        # Get plugin for this capability
        plugin = self.plugin_manager.get_plugin(capability_name)
        
        if plugin is None:
            error_data = {
                "error": f"Unknown capability: {capability_name}",
                "status": "failed"
            }
            return json.dumps(error_data)
        
        try:
            # Parse input parameters
            try:
                params = json.loads(msg.content) if msg.content else {}
            except:
                params = {}
            
            # Execute plugin (pass self as worker_sdk for worker-to-worker calls)
            result = plugin.execute(params, worker_sdk=self)
            
            return json.dumps(result)
            
        except Exception as e:
            print(f"  ✗ Plugin execution error: {e}")
            traceback.print_exc()
            
            error_data = {
                "error": str(e),
                "status": "failed",
                "capability": capability_name
            }
            return json.dumps(error_data)
    
    def connect_and_run(self):
        """Connect to Hub and start processing messages"""
        # Load plugins first
        self.load_plugins()
        
        try:
            print(f"Connecting to gRPC Hub...")
            self.channel = grpc.insecure_channel(self.hub_address)
            grpc.channel_ready_future(self.channel).result(timeout=10)
            print(f"✓ Connected to Hub")
            
            self.stub = hub_pb2_grpc.HubServiceStub(self.channel)
            
            # Create message queue
            send_queue = queue.Queue()
            self.send_queue = send_queue
            
            # Generator function for sending messages
            def request_generator():
                try:
                    # Send registration
                    capabilities = self.plugin_manager.get_all_capabilities()
                    
                    registration_data = {
                        "worker_id": self.worker_id,
                        "worker_type": "python",
                        "capabilities": capabilities,
                        "metadata": {
                            "version": "2.0.0",
                            "description": "Python worker with plugin system",
                            "plugin_count": len(self.plugins)
                        }
                    }
                    
                    register_msg = hub_pb2.Message(
                        id=f"register-{int(time.time() * 1000000)}",
                        to="hub",
                        channel="system",
                        content=json.dumps(registration_data),
                        timestamp=datetime.now().isoformat(),
                        type=hub_pb2.REGISTER
                    )
                    setattr(register_msg, 'from', self.worker_id)
                    yield register_msg
                    print(f"📤 Sent registration with {len(capabilities)} capabilities")
                    
                    # Keep generator alive
                    while self.running:
                        try:
                            msg = send_queue.get(timeout=1)
                            yield msg
                        except queue.Empty:
                            continue
                            
                except Exception as e:
                    print(f"Generator error: {e}")
                    traceback.print_exc()
            
            # Start bidirectional streaming
            print(f"📡 Starting bidirectional stream...")
            response_stream = self.stub.Connect(request_generator())
            
            print(f"✓ Registered with Hub as '{self.worker_id}'")
            print(f"📨 Listening for requests with {len(self.plugins)} plugins loaded...\n")
            
            self.running = True
            
            # Receive loop
            def receive_loop():
                try:
                    for msg in response_stream:
                        if not self.running:
                            break
                        
                        try:
                            msg_from = getattr(msg, 'from')
                            msg_type = msg.type
                            
                            print(f"📬 Received message:")
                            print(f"   From: {msg_from}, Type: {msg_type}, Channel: {msg.channel}")
                            
                            # Handle different message types
                            if msg_type == hub_pb2.RESPONSE:
                                self._handle_worker_call_response(msg)
                                continue
                            
                            elif msg_type == hub_pb2.WORKER_CALL:
                                # Another worker calling us
                                print(f"   → Worker-to-worker call")
                                response_content = self.process_message(msg)
                                
                                response_msg = hub_pb2.Message(
                                    id=f"resp-{int(time.time() * 1000000)}",
                                    to=msg_from,
                                    channel=msg.channel,
                                    content=response_content,
                                    timestamp=datetime.now().isoformat(),
                                    type=hub_pb2.RESPONSE
                                )
                                setattr(response_msg, 'from', self.worker_id)
                                response_msg.metadata['request_id'] = msg.id
                                send_queue.put(response_msg)
                                print(f"   ✓ Queued response\n")
                            
                            else:
                                # Regular REQUEST
                                print(f"   → Regular request")
                                response_content = self.process_message(msg)
                                
                                response_msg = hub_pb2.Message(
                                    id=f"resp-{int(time.time() * 1000000)}",
                                    to=msg_from,
                                    channel=msg.channel,
                                    content=response_content,
                                    timestamp=datetime.now().isoformat(),
                                    type=hub_pb2.RESPONSE
                                )
                                setattr(response_msg, 'from', self.worker_id)
                                send_queue.put(response_msg)
                                print(f"   ✓ Queued response\n")
                        
                        except Exception as e:
                            print(f"✗ Error processing message: {e}")
                            traceback.print_exc()
                
                except grpc.RpcError as e:
                    if e.code() != grpc.StatusCode.CANCELLED:
                        print(f"✗ gRPC Error: {e.code()} - {e.details()}")
                except Exception as e:
                    print(f"✗ Receive loop error: {e}")
                    traceback.print_exc()
                finally:
                    self.running = False
            
            # Start receive thread
            receive_thread = threading.Thread(target=receive_loop, daemon=True)
            receive_thread.start()
            
            # Keep main thread alive
            print("Worker running. Press Ctrl+C to stop.\n")
            while self.running:
                time.sleep(1)
        
        except grpc.RpcError as e:
            print(f"\n✗ gRPC Error: {e.code()} - {e.details()}")
        except Exception as e:
            print(f"\n✗ Connection error: {e}")
            traceback.print_exc()
        finally:
            self.running = False
            if self.channel:
                self.channel.close()
            self.plugin_manager.unload_all_plugins()


def main():
    print("🐍 Python Worker (Plugin System)")
    print("=" * 50)
    
    worker_id = os.getenv('WORKER_ID', 'python-worker')
    hub_address = os.getenv('HUB_ADDRESS', 'localhost:50051')
    
    print(f"Worker ID: {worker_id}")
    print(f"Hub Address: {hub_address}")
    print("=" * 50)
    
    worker = PluginWorker(worker_id, hub_address)
    
    try:
        worker.connect_and_run()
    except KeyboardInterrupt:
        print("\n\n✗ Shutting down...")
        worker.running = False


if __name__ == '__main__':
    main()

File: services/node-worker/worker-plugin-system.js
#!/usr/bin/env node

const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const path = require('path');
const fs = require('fs');

const PluginManager = require('./plugins/plugin-manager');

// Load proto
const PROTO_PATH = path.join(__dirname, '../proto/hub.proto');
const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
    keepCase: true,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true
});
const hubProto = grpc.loadPackageDefinition(packageDefinition).hub;

/**
 * Node.js Worker with Plugin System
 */
class NodeWorker {
    constructor(workerId, hubAddress) {
        this.workerId = workerId;
        this.hubAddress = hubAddress;
        this.pluginManager = new PluginManager();
        this.stream = null;
        this.pendingCalls = new Map();
        this.connected = false;
    }

    /**
     * Initialize worker
     */
    async initialize() {
        console.log('🟢 Node.js Worker Starting...');
        console.log(`   Worker ID: ${this.workerId}`);
        console.log(`   Hub Address: ${this.hubAddress}`);
        
        // Discover and load all plugins
        console.log('\n📦 Loading plugins...');
        this.pluginManager.discoverPlugins();
        
        // Also load worker-to-worker plugins
        this.loadWorkerToWorkerPlugins();
        
        const capabilities = this.pluginManager.getCapabilities();
        console.log(`\n✅ Registered ${capabilities.length} capabilities:`);
        capabilities.forEach(cap => {
            console.log(`   - ${cap.name}: ${cap.description}`);
        });
    }

    /**
     * Load worker-to-worker communication plugins
     */
    loadWorkerToWorkerPlugins() {
        const workerToWorkerDir = path.join(__dirname, 'worker-to-worker');
        
        if (!fs.existsSync(workerToWorkerDir)) {
            return;
        }

        console.log(`\n🔄 Loading worker-to-worker plugins...`);
        const files = fs.readdirSync(workerToWorkerDir);
        
        for (const file of files) {
            if (file.endsWith('-plugin.js')) {
                try {
                    const PluginClass = require(path.join(workerToWorkerDir, file));
                    const plugin = new PluginClass();
                    const pluginName = plugin.getName();
                    
                    this.pluginManager.plugins.set(pluginName, plugin);
                    console.log(`   ✓ Loaded: ${pluginName} (${file})`);
                } catch (error) {
                    console.error(`   ✗ Failed to load ${file}:`, error.message);
                }
            }
        }
    }

    /**
     * Connect to Hub and establish bidirectional stream
     */
    connect() {
        const client = new hubProto.HubService(
            this.hubAddress,
            grpc.credentials.createInsecure()
        );

        this.stream = client.Connect();

        // Handle incoming messages from Hub
        this.stream.on('data', (message) => {
            this.handleHubMessage(message);
        });

        this.stream.on('error', (error) => {
            console.error('❌ Stream error:', error.message);
            this.connected = false;
            setTimeout(() => this.reconnect(), 5000);
        });

        this.stream.on('end', () => {
            console.log('⚠️  Stream ended');
            this.connected = false;
            setTimeout(() => this.reconnect(), 5000);
        });

        // Register worker with capabilities
        this.register();
    }

    /**
     * Register worker with Hub
     */
    register() {
        const capabilities = this.pluginManager.getCapabilities();
        
        const registrationData = {
            worker_id: this.workerId,
            worker_type: 'node',
            capabilities: capabilities,
            metadata: {
                version: '1.0.0',
                description: 'Node.js worker with plugin system',
                plugin_count: capabilities.length
            }
        };

        const registerMessage = {
            type: 3, // REGISTER = 3
            from: this.workerId,
            to: 'hub',
            content: JSON.stringify(registrationData),
            action: 'register'
        };

        this.stream.write(registerMessage);
        this.connected = true;
        console.log('\n✅ Worker registered with Hub');
        
        // Send heartbeat every 30 seconds
        this.startHeartbeat();
    }

    /**
     * Start heartbeat to keep connection alive
     */
    startHeartbeat() {
        setInterval(() => {
            if (this.connected && this.stream) {
                this.stream.write({
                    type: 'HEARTBEAT',
                    from: this.workerId,
                    heartbeat: { timestamp: new Date().toISOString() }
                });
            }
        }, 30000);
    }

    /**
     * Handle incoming messages from Hub
     */
    async handleHubMessage(message) {
        // MessageType enum: DIRECT=0, BROADCAST=1, CHANNEL=2, REGISTER=3, REQUEST=4, RESPONSE=5, WORKER_CALL=6
        // Note: gRPC may send as number OR string depending on serialization
        
        try {
            // Normalize type to string for comparison
            const messageType = typeof message.type === 'string' ? message.type : String(message.type);
            
            switch (messageType) {
                case '4':
                case 'REQUEST':
                    await this.handleRequest(message);
                    break;
                    
                case '5':
                case 'RESPONSE':
                    this.handleResponse(message);
                    break;
                    
                case '3':
                case 'REGISTER':
                    console.log('✓ Registration acknowledged');
                    break;
                    
                default:
                    // Silently ignore heartbeats and unknown types
                    break;
            }
        } catch (error) {
            console.error('❌ Error handling message:', error.message);
        }
    }

    /**
     * Handle capability request
     */
    async handleRequest(message) {
        // Extract request info from message fields (not content)
        const request_id = message.id;  // Request ID is in message.id
        const original_sender = message.from;  // Who sent the request (web-api)
        const capability = message.metadata?.capability || message.channel;  // Capability in metadata or channel
        
        // Parse user data from message.content
        let params = {};
        if (message.content && message.content.trim() !== '' && message.content !== '{}') {
            try {
                params = JSON.parse(message.content);
            } catch (e) {
                console.error('❌ Failed to parse content:', e.message);
                return;
            }
        }
        
        console.log(`📥 Request: ${capability} (ID: ${request_id})`);
        
        try {
            // Get plugin
            const plugin = this.pluginManager.getPlugin(capability);
            if (!plugin) {
                throw new Error(`Unknown capability: ${capability}`);
            }

            // Execute plugin with context
            const context = {
                workerId: this.workerId,
                callWorker: this.callWorker.bind(this)
            };
            
            const result = await plugin.execute(params, context);

            // Send response back to original sender
            this.sendResponse(request_id, original_sender, result, 'success');
            console.log(`✅ Response sent for: ${capability}`);
            
        } catch (error) {
            console.error(`❌ Error processing ${capability}:`, error.message);
            this.sendResponse(request_id, original_sender, { error: error.message }, 'error');
        }
    }

    /**
     * Send response to Hub
     */
    sendResponse(requestId, targetClient, result, status) {
        // Send response back - the result data goes directly in content as JSON
        const responseMessage = {
            id: requestId,  // Same ID as request
            type: 5, // RESPONSE = 5
            from: this.workerId,
            to: targetClient,  // Send back to original requester (web-api)
            content: JSON.stringify(result),  // Just the result data
            timestamp: new Date().toISOString()
        };

        this.stream.write(responseMessage);
    }

    /**
     * Call another worker's capability
     */
    callWorker(targetWorkerId, capability, params, timeout = 30000) {
        return new Promise((resolve, reject) => {
            const requestId = `${this.workerId}-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
            
            // Store pending call
            const timer = setTimeout(() => {
                this.pendingCalls.delete(requestId);
                reject(new Error(`Timeout calling ${targetWorkerId}.${capability}`));
            }, timeout);

            this.pendingCalls.set(requestId, {
                resolve,
                reject,
                timer,
                targetWorkerId,
                capability
            });

            // Send WORKER_CALL message to Hub
            const callMessage = {
                type: 'WORKER_CALL',
                from: this.workerId,
                worker_call: {
                    request_id: requestId,
                    target_worker_id: targetWorkerId,
                    capability: capability,
                    data: JSON.stringify(params)
                }
            };

            this.stream.write(callMessage);
        });
    }

    /**
     * Handle response from another worker
     */
    handleResponse(message) {
        const { request_id, status, data } = message.response;
        
        const pending = this.pendingCalls.get(request_id);
        if (!pending) {
            return;
        }

        clearTimeout(pending.timer);
        this.pendingCalls.delete(request_id);

        try {
            const result = JSON.parse(data);
            
            if (status === 'success') {
                pending.resolve(result);
            } else {
                pending.reject(new Error(result.error || 'Worker call failed'));
            }
        } catch (error) {
            pending.reject(error);
        }
    }

    /**
     * Reconnect to Hub
     */
    reconnect() {
        console.log('🔄 Reconnecting to Hub...');
        this.connect();
    }

    /**
     * Start the worker
     */
    async start() {
        await this.initialize();
        this.connect();
        
        console.log('\n🚀 Node.js Worker is running!\n');
    }
}

// Main execution
const workerId = process.env.WORKER_ID || 'node-worker';
const hubAddress = process.env.HUB_ADDRESS || 'localhost:50051';

const worker = new NodeWorker(workerId, hubAddress);
worker.start().catch(error => {
    console.error('❌ Fatal error:', error);
    process.exit(1);
});

// Graceful shutdown
process.on('SIGINT', () => {
    console.log('\n👋 Shutting down gracefully...');
    process.exit(0);
});

process.on('SIGTERM', () => {
    console.log('\n👋 Shutting down gracefully...');
    process.exit(0);
});
